kernel mode: validate conflicting routes

cathay4t opened the pull request:
Kernel does not allows desired multiple routes with the same destination
and metric.
When user desire so, it is unclear whether user made a fault or want
ECMP routes.

In NM mode, that YAML will generate ECMP routes for IPv6 and append
desired routes before existing for IPv4. This patch does not change that
because it might break working YAMLs.

For kernel mode, this patch raise better error instead of kernel error
code `-22`:

    NmstateError: InvalidArgument: Multiple routes to 2001:db8:2::/64 are
    sharing the same metric without weight, please neither use `state:
    absent` to remove others or define `weight` for ECMP.

Integration test case included.

Resolves: https://issues.redhat.com/browse/RHEL-54558

jvaclav-rh commented on rust/src/lib/nispor/route.rs:
```diff
@@ -289,9 +291,53 @@ fn nmstate_to_nispor_route_conf(
+            // The `merged_routes.merged` does not have absent route.
+            hashed_rts
+                .entry((
+                    rt.destination.clone().unwrap_or_default(),
```
Here, it would probably be better to expect `rt.destination` to be defined, and raise error if not. Since we are reading merged routes, if it is None, something must be very wrong. Similar to: https://github.com/nmstate/nmstate/commit/48b139c87f9d9b2f248e27e6b8aa964b483ed176#diff-1f2a539da936256735b6b03fff9470b4ca657f83abfb6370dd94b27cdeb4e222R889-R900. 

This also avoids cloning it for every route, and you can use `&str` as part of the key.

cathay4t replied:
Yes. I should use reference here.

jvaclav-rh commented on rust/src/lib/nispor/route.rs:
```diff
@@ -289,9 +291,53 @@ fn nmstate_to_nispor_route_conf(
+                    ErrorKind::InvalidArgument,
+                    format!(
+                        "Multiple routes to {dst} are sharing the same metric \
+                         without weight defined, please neither use `state: \
+                         absent` to remove others or define `weight` for ECMP."
```
I think we shouldn't tell the user to define `weight`, since kernel mode does not support this, and this error is only raised in kernel mode.

cathay4t replied:
Sure.

jvaclav-rh commented on rust/src/lib/nispor/route.rs:
```diff
@@ -289,9 +291,58 @@ fn nmstate_to_nispor_route_conf(
+            hashed_rts
+                .entry((dst, rt.metric.and_then(|m| u32::try_from(m).ok())))
+                .or_default()
+                .push(rt);
```
Instead of inserting into a vector here, we could save just `&RouteEntry` into the map, and use the return value of HashMap::Insert(), which returns `Some(old_value)` if the key is occupied already - which would indicate conflicting route.

The `weight` check from below could also moved into this loop, it would save us from having to iterate over the same data twice.

cathay4t replied:
Good point!

jvaclav-rh commented on rust/src/lib/nispor/route.rs:
```diff
@@ -289,9 +291,57 @@ fn nmstate_to_nispor_route_conf(
+                    ErrorKind::InvalidArgument,
+                    format!(
+                        "Multiple routes to {dst} are sharing the same metric \
+                         without weight defined, please neither use `state: \
```
```suggestion
                         without weight defined, please use `state: \
```

cathay4t replied:
Fixed. Thanks!

ihuguet commented on rust/src/lib/nispor/route.rs:
```diff
@@ -289,9 +291,57 @@ fn nmstate_to_nispor_route_conf(
+                return Err(NmstateError::new(
+                    ErrorKind::InvalidArgument,
+                    format!(
+                        "Multiple routes to {dst} are sharing the same metric \
+                         without weight defined, please neither use `state: \
+                         absent` to remove others."
```
This error message is missleading too, as it mentions weight which is not supported in kernel mode. The word "neither" is wrong there too.

Suggestion: Multiple routes to {dst} with the same metric. Use `state:  absent` to remove previous routes.

cathay4t replied:
Fixed. Thanks!

ihuguet commented on rust/src/lib/nispor/route.rs:
```diff
@@ -289,9 +291,57 @@ fn nmstate_to_nispor_route_conf(
+            };
+
+            if hashed_rts
+                .insert(
+                    (dst, rt.metric.and_then(|m| u32::try_from(m).ok())),
+                    rt,
+                )
```
If `rt.metric` is None, routes with same dst and no metric will match. This is intended, right?

cathay4t replied:
Yes

ihuguet commented on tests/integration/route_test.py:
```diff
@@ -2505,3 +2505,51 @@ def test_apply_ignored_routes_to_iface(eth1_static_ip):
+                prefix-length: 64
+        routes:
+         config:
+           - destination: 203.0.113.0/24
+             next-hop-address: 192.0.2.1
+             next-hop-interface: veth1
+             metric: 109
+           - destination: 203.0.113.0/24
+             next-hop-address: 192.0.2.2
+             next-hop-interface: veth1
+             metric: 109
+           - destination: 2001:db8:2::/64
+             next-hop-address: 2001:db8:1::2
+             next-hop-interface: veth1
+             metric: 102
+           - destination: 2001:db8:2::/64
+             next-hop-address: 2001:db8:1::3
+             next-hop-interface: veth1
+             metric: 102
```
Doing all at the same time you won't know if it works well both for IPv4 and IPv6, or only for one of them. I would do them in separate steps.

I'd also add one step with routes without metric. And another with one with metric and other without metric, that they shouldn't collide.

cathay4t replied:
New patch uploaded with `pytest.mark.parametrize` covering all 4 cases.

cathay4t committed:
kernel mode: validate conflicting routes

Kernel does not allows desired multiple routes with the same destination
and metric.
When user desire so, it is unclear whether user made a fault or want
ECMP routes.

In NM mode, that YAML will generate ECMP routes for IPv6 and append
desired routes before existing for IPv4. This patch does not change that
because it might break working YAMLs.

For kernel mode, this patch raise better error instead of kernel error
code `-22`:

    NmstateError: InvalidArgument: Multiple routes to 2001:db8:2::/64 are
    sharing the same metric without weight, please neither use `state:
    absent` to remove others.

Integration test case included.

Resolves: https://issues.redhat.com/browse/RHEL-54558

Signed-off-by: Gris Ge <fge@redhat.com>
